controller/snapshot_controller.go

```diff
@@ -183,6 +189,35 @@ func filterSnapshotsForEngineEnqueuing(oldEngine, curEngine *longhorn.Engine, sn
 	return targetSnapshots
 }
 
+func (sc *SnapshotController) enqueueVolumeChange(obj interface{}) {
+	vol, ok := obj.(*longhorn.Volume)
+	if !ok {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("received unexpected obj: %#v", obj))
+			return
+		}
+		// use the last known state, to enqueue, dependent objects
+		vol, ok = deletedState.Obj.(*longhorn.Volume)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("DeletedFinalStateUnknown contained invalid object: %#v", deletedState.Obj))
+			return
+		}
+	}
+	if vol.DeletionTimestamp.IsZero() {
+		return
+	}
+
+	snapshots, err := sc.ds.ListVolumeSnapshotsRO(vol.Name)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("snapshot controller failed to list snapshots when enqueuing volume %v: %v", vol.Name, err))
+		return
```

Comment:
Is it possible to cause orphaned snapshots here? if just somehow it's failed to get snapshots, and the volume has been deleted asynchronously?

Reply:
After discussing with @shuo-wu , this is possible but right now it's unable to handle via the current asynchronous mechanism (controller) except by leveraging the orphan mechanism.

This can be considered later, rather low priority.